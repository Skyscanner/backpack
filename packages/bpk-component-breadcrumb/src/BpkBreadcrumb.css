@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-breadcrumb{display:flex;margin:0;padding:0;flex-wrap:wrap;list-style-type:none}
