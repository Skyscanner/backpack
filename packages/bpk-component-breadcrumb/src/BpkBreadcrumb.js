/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import PropTypes from 'prop-types';
import React, { type Node } from 'react';

import { cssModules } from '../../bpk-react-utils';

import STYLES from './BpkBreadcrumb.module.scss';

const getClassName = cssModules(STYLES);

type SchemaMetaDataItem = {
  url: string,
  label: string,
};

export type Props = {
  children: Node,
  schemaMetaData: ?(SchemaMetaDataItem[]),
  label: string,
};

/*
  The google structured data reference for the stringified output of
  the following function is here:
  https://developers.google.com/search/docs/data-types/breadcrumb
*/
const buildMetaData = (schemaMetaData: SchemaMetaDataItem[]): string => {
  const itemListElement = schemaMetaData.map((schemaMetaDataItem, index) => ({
    '@type': 'ListItem',
    position: index + 1,
    item: {
      '@id': schemaMetaDataItem.url,
      name: schemaMetaDataItem.label,
    },
  }));

  return JSON.stringify({
    '@context': 'http://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement,
  });
};

class BpkBreadcrumb extends React.Component<Props> {
  metaData: ?string;

  static defaultProps = {
    schemaMetaData: null,
  };

  constructor(props: Props) {
    super(props);

    this.metaData = props.schemaMetaData && buildMetaData(props.schemaMetaData);
  }

  render() {
    const { children, label, schemaMetaData, ...rest } = this.props;

    return (
      <React.Fragment>
        {this.metaData && (
          <script
            type="application/ld+json"
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{ __html: this.metaData }}
          />
        )}
        {/* $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'. */}
        <nav aria-label={label} {...rest}>
          <ol className={getClassName('bpk-breadcrumb')}>{children}</ol>
        </nav>
      </React.Fragment>
    );
  }
}

BpkBreadcrumb.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.string.isRequired,
  schemaMetaData: PropTypes.arrayOf(
    PropTypes.shape({
      url: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    }),
  ),
};

BpkBreadcrumb.defaultProps = {
  schemaMetaData: null,
};

export default BpkBreadcrumb;
