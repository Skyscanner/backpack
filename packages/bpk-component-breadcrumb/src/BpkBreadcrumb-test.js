/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import React from 'react';
import renderer from 'react-test-renderer';

import BpkBreadcrumb from './BpkBreadcrumb';

describe('BpkBreadcrumb', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(
        <BpkBreadcrumb label="My breadcrumbs">
          <div>Anything can go in here</div>
        </BpkBreadcrumb>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a custom class name', () => {
    const tree = renderer
      .create(
        <BpkBreadcrumb label="My breadcrumbs" className="my-custom-class">
          <div>Anything can go in here</div>
        </BpkBreadcrumb>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with arbitrary props', () => {
    const tree = renderer
      .create(
        <BpkBreadcrumb
          label="My breadcrumbs"
          testID="arbitrary value" // <-- arbitrary prop
        >
          <div>Anything can go in here</div>
        </BpkBreadcrumb>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
