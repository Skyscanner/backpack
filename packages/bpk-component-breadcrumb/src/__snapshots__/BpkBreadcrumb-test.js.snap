// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkBreadcrumb should render correctly 1`] = `
<nav
  aria-label="My breadcrumbs"
>
  <ol
    className="bpk-breadcrumb"
  >
    <div>
      Anything can go in here
    </div>
  </ol>
</nav>
`;

exports[`BpkBreadcrumb should render correctly with a custom class name 1`] = `
<nav
  aria-label="My breadcrumbs"
  className="my-custom-class"
>
  <ol
    className="bpk-breadcrumb"
  >
    <div>
      Anything can go in here
    </div>
  </ol>
</nav>
`;

exports[`BpkBreadcrumb should render correctly with arbitrary props 1`] = `
<nav
  aria-label="My breadcrumbs"
  testID="arbitrary value"
>
  <ol
    className="bpk-breadcrumb"
  >
    <div>
      Anything can go in here
    </div>
  </ol>
</nav>
`;

exports[`BpkBreadcrumb should render correctly with schema meta data 1`] = `
Array [
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"@context\\":\\"http://schema.org\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"item\\":{\\"@id\\":\\"http://www.skyscanner.net\\",\\"name\\":\\"home\\"}},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"item\\":{\\"@id\\":\\"http://www.skyscanner.net/hotels\\",\\"name\\":\\"hotels\\"}}]}",
      }
    }
    type="application/ld+json"
  />,
  <nav
    aria-label="My breadcrumbs"
  >
    <ol
      className="bpk-breadcrumb"
    >
      <div>
        Anything can go in here
      </div>
    </ol>
  </nav>,
]
`;
